### PR TITLE
Add CentOS support

### PR DESCRIPTION
**Pull request by naamunds**
The CoreFx build using build.sh is currently broken on CentOS because it
doesn't restore the right version of the .NET CLI.  This change fixes
that and also makes a few other minor edits:

- Remove ToolNugetRuntimeId, which is no longer used.
- Don't use "sh" when invoking init-tools.sh, because Bash is needed to
use the "source" command.
- Add a few more console output messages to init-tools.sh, similar to
the ones already in init-tools.cmd.
- Fix indentation in init-tools.sh.
- Set the default TestNugetRuntimeId in build.sh when the target OS is
specified to be FreeBSD, NetBSD, or Windows.

This fix is also dependent on https://github.com/dotnet/buildtools/pull/435

See https://github.com/dotnet/corefx/issues/4650

cc: @ellismg @joperezr @eerhardt

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -217,7 +216,15 @@ case $OSName in
 
     Linux)
         __HostOS=Linux
-        __TestNugetRuntimeId=ubuntu.14.04-x64
+        source /etc/os-release
+        if [ $ID = "centos" ]; then
+            __TestNugetRuntimeId=centos.7.1-x64
```
Comment (ellismg) [inline build.sh]: Can you change this to `centos.7-x64`?  That the RID that will be generated by default on CentOS platforms.  I would like to remove the centos.7.1-x64 rid at some point because it's meaningless.

File: src/Common/test-runtime/project.json [hunk 2]
```diff
@@ -22,6 +22,7 @@
     "win7-x64": {},
     "win7-x86": {},
     "ubuntu.14.04-x64": {},
-    "osx.10.10-x64": {}
+    "osx.10.10-x64": {},
+    "centos.7.1-x64": {}
```
Comment (ellismg) [inline src/Common/test-runtime/project.json]: Update to `centos.7-x64` if you change it everywhere else.

File: init-tools.sh [hunk 3]
```diff
@@ -22,47 +22,59 @@ case $OSName in
 
     Linux)
         OS=Linux
-        __DOTNET_PKG=dotnet-ubuntu-x64
+        source /etc/os-release
+        if [ $ID = "centos" ]; then
```
Comment (joperezr) [inline init-tools.sh]: Shouldn't this require '==' instead of just '='? Same for all other comparisons.

File: init-tools.sh [hunk 4]
```diff
@@ -22,47 +22,59 @@ case $OSName in
...
+            __DOTNET_PKG=dotnet-centos-x64
+        elif [ $ID = "ubuntu" -o $ID = "debian" ]; then
+            __DOTNET_PKG=dotnet-ubuntu-x64
+        else
+            echo "Unsupported ID '$ID' detected. Downloading ubuntu-x64 tools."
```
Comment (joperezr) [inline init-tools.sh]: I'm not sure if 'ID' is understandable. How about 'Unsuported Linux Distribution ...' or something like that?

File: init-tools.sh [hunk 5]
```diff
@@ -22,47 +22,59 @@ case $OSName in
...
-        echo "Unsupported OS $OSName detected. Downloading ubuntu-x64 tools"
+        echo "Unsupported OS '$OSName' detected. Downloading ubuntu-x64 tools."
         OS=Linux
         __DOTNET_PKG=dotnet-ubuntu-x64
         ;;
 esac
 
 if [ ! -e $__PROJECT_JSON_FILE ]; then
- if [ -e $__TOOLRUNTIME_DIR ]; then rm -rf -- $__TOOLRUNTIME_DIR; fi
+    if [ -e $__TOOLRUNTIME_DIR ]; then rm -rf -- $__TOOLRUNTIME_DIR; fi
```
Comment (joperezr) [inline init-tools.sh]: is the indenting an intended change?

File: init-tools.sh [hunk 6]
```diff
@@ -22,47 +22,59 @@ case $OSName in
...
+    if [ ! -e $__BUILD_TOOLS_PATH ]; then
+        echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
+        $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE
+    fi
 
- sh $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR
- chmod a+x $__TOOLRUNTIME_DIR/corerun
+    echo "Initializing BuildTools..."
+    $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR
+    chmod a+x $__TOOLRUNTIME_DIR/corerun
```
Comment (joperezr) [inline init-tools.sh]: This is actually done on the buildtools side, so no need of running this again.

File: init-tools.sh [hunk 7]
```diff
@@ -22,47 +22,59 @@ case $OSName in
...
+        # curl has HTTPS CA trust-issues less often than wget, so lets try that first.
+        which curl > /dev/null 2> /dev/null
+        if [ $? -ne 0 ]; then
+            mkdir -p "$__DOTNET_PATH"
+            wget -q -O $__DOTNET_PATH/dotnet.tar https://dotnetcli.blob.core.windows.net/dotnet/dev/Binaries/${__DOTNET_TOOLS_VERSION}/${__DOTNET_PKG}.${__DOTNET_TOOLS_VERSION}.tar.gz
+        else
+            curl -sSL --create-dirs -o $__DOTNET_PATH/dotnet.tar https://dotnetcli.blob.core.windows.net/dotnet/dev/Binaries/${__DOTNET_TOOLS_VERSION}/${__DOTNET_PKG}.${__DOTNET_TOOLS_VERSION}.tar.gz
+        fi
+        cd $__DOTNET_PATH
+        tar -xf $__DOTNET_PATH/dotnet.tar
```
Comment (joperezr) [inline init-tools.sh]: Given that you are already doing the echo to mirror what we have in the cmd script, could we disable the verbosity of the tar command?

File: init-tools.sh [hunk 8]
```diff
@@ -22,47 +22,59 @@ case $OSName in
...
- echo $__PROJECT_JSON_CONTENTS > "$__PROJECT_JSON_FILE"
+    mkdir "$__PROJECT_JSON_PATH"
+    echo $__PROJECT_JSON_CONTENTS > "$__PROJECT_JSON_FILE"
 
- if [ ! -e $__BUILD_TOOLS_PATH ]; then
-    $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE
- fi
+    if [ ! -e $__BUILD_TOOLS_PATH ]; then
+        echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
+        $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE
```
Comment (joperezr) [inline init-tools.sh]: Can we send the output of this command to a init-tools.log to mimic the cmd?

**Comment by joperezr** [comment]
@naamunds I added some comments about making this script mimic the output of the cmd equivalent, which basically outputs all the verbosity to a log. All of those comments are not related nor required for your PR, so if you want I can log an issue and assign it to myself and do that work later.

File: init-tools.sh (hunk 3 above)
Comment (naamunds) [inline init-tools.sh]: According to http://www.tldp.org/LDP/abs/html/comparison-ops.html, `==` and `=` are equivalent for string comparison in Bash, although only `=` is supported in SH. I think it does make sense to use `==` since this is exclusively a Bash script.

File: init-tools.sh (hunk 4 above)
Comment (naamunds) [inline init-tools.sh]: Good point. I've made that change.

File: init-tools.sh (hunk 5 above)
Comment (naamunds) [inline init-tools.sh]: Yes; the indentation in this file is inconsistent, so this change fixes it.

File: init-tools.sh (hunk 7 above)
Comment (naamunds) [inline init-tools.sh]: I had noticed that earlier and was planning to fix it, but it was actually fixed in https://github.com/dotnet/corefx/commit/c180cc0ef7adfde965514e0bce85e09668b09a9d.

File: init-tools.sh (hunk 6 above)
Comment (naamunds) [inline init-tools.sh]: That's true. I didn't add it with this change, but I agree that it makes sense to remove it, so I've done so.

File: init-tools.sh (hunk 8 above)
Comment (naamunds) [inline init-tools.sh]: That's a good suggestion, but I'd prefer that to be done in a separate PR. If you want to log an issue for it, that would be great.

File: build.sh (hunk 1 above)
Comment (naamunds) [inline build.sh]: Fixed.

File: src/Common/test-runtime/project.json (hunk 2 above)
Comment (naamunds) [inline src/Common/test-runtime/project.json]: Fixed.

File: src/Common/test-runtime/project.json [hunk 9]
```diff
@@ -22,6 +22,8 @@
     "win7-x64": {},
     "win7-x86": {},
     "ubuntu.14.04-x64": {},
-    "osx.10.10-x64": {}
+    "osx.10.10-x64": {},
+    "centos.7-x64": {},
+    "debian.8.2-x64": {}
```
Comment (ellismg) [inline src/Common/test-runtime/project.json]: Need `rhel.7-x64` here as well.

**Comment by ellismg** [comment]
One comment, LGTM otherwise.

File: src/Common/test-runtime/project.json (hunk 9 above)
Comment (naamunds) [inline src/Common/test-runtime/project.json]: Fixed.

**Comment by joperezr** [comment]
LGTM
